# COVID-19 final project
This repo contains all files related to workflow used in the project:
-Code.cwl > the source code to be implemented in the cloud platform.
-SRA_meta.txt > the file used to fetch all data in one execution turn.
-SupplementaryFile.xlsx > the detailed result for each sample.
-UShER.json > the file to visualize the phylogenetics tree.
-UShER_meta.tsv > the additional metadata in visualizing the phylogenetics tree.
-VirusViz.json.gz > the file to visualize the variant comparison.
-viral.tar > the reference index used in the project

## How to Use
1. Login to [Cancer Genomics Cloud](https://cgc-accounts.sbgenomics.com/auth/login) platform.
2. Enter your own **Projects**.
3. On the top left menu, select **Apps** option.
4. Click **Add app** button on the top right corner.
5. Select **Create New App**,then click **Create a Workflow** button.
6. Type the workflow name, then click **Create** button.
7. Select **Code** to change the layout view.
8. Replace all appeared default code with our workflow code provided in this page called *Code.cwl*.
9. Click **Save and Run** button on the top right corner to make the workflow available in your own project.
